gh-117233: Detect libcrypto BLAKE2, Shake, SHA3, and Truncated-SHA512 support at hashlib build time

WillChilds-Klein opened the pull request:
# Notes

While OpenSSL supports both "b" and "s" variants of the BLAKE2 hash
function, other cryptographic libraries may lack support for one or both
of the variants. This commit modifies `hashlib`'s C code to detect
whether or not the linked libcrypto supports each BLAKE2 variant, and
elides references to each variant's NID accordingly. In cases where the
underlying libcrypto doesn't fully support BLAKE2, CPython's
`./configure` script can be given the following flag to use CPython's
interned BLAKE2 implementation: `--with-builtin-hashlib-hashes=blake2`.

This pull request implements Issue #117233.

<!--
Thanks for your contribution!
Please read this comment in its entirety. It's quite important.

# Pull Request title

It should be in the following format:

```
gh-NNNNN: Summary of the changes made
```

Where: gh-NNNNN refers to the GitHub issue number.

Most PRs will require an issue number. Trivial changes, like fixing a typo, do not need an issue.

# Backport Pull Request title

If this is a backport PR (PR made against branches other than `main`),
please ensure that the PR title is in the following format:

```
[X.Y] <title from the original PR> (GH-NNNN)
```

Where: [X.Y] is the branch name, e.g. [3.6].

GH-NNNN refers to the PR number from `main`.

-->


<!-- gh-issue-number: gh-117233 -->
* Issue: gh-117233
<!-- /gh-issue-number -->

WillChilds-Klein committed:
Detect crypto library BLAKE2 support

While OpenSSL supports both "b" and "s" variants of the BLAKE2 hash
function, other cryptographic libraries may lack support for one or both
of the variants. This commit modifies `hashlib`'s C code to detect
whether or not the linked libcrypto supports each BLAKE2 variant, and
elides references to each variant's NID accordingly. In cases where the
underlying libcrypto doesn't fully support BLAKE2, CPython's
`./configure` script can be given the following flag to use CPython's
interned BLAKE2 implementation: `--with-builtin-hashlib-hashes=blake2`.

blurb-it[bot] committed:
📜🤖 Added by blurb_it.

WillChilds-Klein commented on Modules/_hashopenssl.c:
```diff
@@ -47,7 +47,9 @@
 #define PY_OPENSSL_HAS_SCRYPT 1
 #define PY_OPENSSL_HAS_SHA3 1
 #define PY_OPENSSL_HAS_SHAKE 1
+#if defined(NID_blake2s256) || defined(NID_blake2b512)
 #define PY_OPENSSL_HAS_BLAKE2 1
+#endif
```
It looks like `PY_OPENSSL_HAS_BLAKE2` [isn't referenced anywhere in CPython](https://github.com/search?q=repo%3Apython%2Fcpython+PY_OPENSSL_HAS_BLAKE2&type=code), so I'm guessing as to what the "correct" behavior here is. Intuitively, I take this symbol to mean "linked libcrypto supports _some_ BLAKE2 functionality". Please correct me if this is wrong.

gpshead replied:
If it builds against both libraries, this will work for now.  What we did in our BoringSSL patch set is blindly disable all of the unsupported hash functions: sha512_224 sha512_256, sha3, shake, and blake2 - per https://github.com/python/cpython/pull/116399/files#diff-b4fd2e5e447a6f18fb2508c350524c5cb1feed8ec92d70822eac54431ef71baf

I like this compile time feature detection idea better.

gpshead replied:
In the update I pushed to the PR I changed the || to && as it is simpler to assume that if a library supports one of these categories it will support them all.  it'd probably cause a build failure if something only supported a fraction of a category - that can be dealt with if it ever arises (doubtful).

gpshead replied:
eh.. || makes sense for blake2 and shake.  reverting that bit of my change. :)

alex reviewed (commented):
Would it make more sense to reference the `OPENSSL_NO_BLAKE2` define, which OpenSSL (and BoringSSL) already expose?

gpshead committed:
Expand build time detection to SHA3, Shake, & truncated SHA512.

Detect BLAKE2, SHA3, Shake, & truncated SHA512 support in the
OpenSSL-ish libcrypto library at build time.  This helps allow hashlib's
\_hashopenssl to be used with libraries that do not to support every
algorithm that upstream OpenSSL does.  Such as AWS-LC & BoringSSL.

gpshead committed:
go back to || for blake2 and shake, remove unused define.

gpshead commented on Lib/test/test_ssl.py:
```diff
@@ -3867,7 +3867,7 @@ def test_min_max_version_mismatch(self):
                 with self.assertRaises(ssl.SSLError) as e:
                     s.connect((HOST, server.port))
-                self.assertRegex("(alert|ALERT)", str(e.exception))
+                self.assertRegex(str(e.exception), "(alert|ALERT)")
```
https://github.com/python/cpython/pull/117351 will take care of this on its own.

WillChilds-Klein replied:
thanks @gpshead.

encukou commented on Modules/_hashopenssl.c:
```diff
@@ -88,22 +94,45 @@ typedef struct {
+#endif
+#if defined(NID_sha512_512)
+# define Py_hash_sha512_256 "sha512_256"
+#endif
```
This typo would disable sha512-256 entirely, wouldn't it?
```suggestion
#if defined(NID_sha512_256)
# define Py_hash_sha512_256 "sha512_256"
#endif
```

gpshead replied:
Thanks for the eyeballs, yes it would have. _(and since that isn't one of our guaranteed available hash functions it would've taken a while to notice)_

gpshead replied:
or realistically, maybe it wouldn't have given which things tend to be defined at the same time.  still, correct change.

gpshead committed:
typo fix, gate sha512-256 on NID_sha512_256

Co-authored-by: Petr Viktorin <encukou@gmail.com>

encukou reviewed (approved):
LGTM now :)
